src/libs/data-table-versions.js

```diff
@@ -89,25 +123,44 @@ export const restoreDataTableVersion = async (workspace, version) => {
   const { workspace: { namespace, name, googleProject, bucketName } } = workspace
 
   const [, objectName] = parseGsUri(version.url)
-  const content = await Ajax().Buckets.getObjectPreview(googleProject, bucketName, objectName, true).then(r => r.text())
-
-  const tableName = tableNameForRestore(version)
-  const file = new File(
-    [_.replace(/^entity:\S+/, `entity:${tableName}_id`, content)],
-    _.last(_.split('/', objectName)),
-    { type: 'text/tab-separated-values' }
-  )
-
-  const filesize = file?.size || Number.MAX_SAFE_INTEGER
-  if (filesize < 524288) { // 512k
-    await Ajax().Workspaces.workspace(namespace, name).importFlexibleEntitiesFileSynchronous(file, { deleteEmptyValues: false })
-    return { tableName, ready: true }
-  } else {
-    const { jobId } = await Ajax().Workspaces.workspace(namespace, name).importFlexibleEntitiesFileAsync(file, { deleteEmptyValues: false })
-    asyncImportJobStore.update(Utils.append({ targetWorkspace: { namespace, name }, jobId }))
-    notifyDataImportProgress(jobId)
-    return { tableName, ready: false }
+  const zipData = await Ajax().Buckets.getObjectPreview(googleProject, bucketName, objectName, true).then(r => r.blob())
+
+  const zip = await JSZip.loadAsync(zipData)
+
+  const restoredTableName = tableNameForRestore(version)
+  const entities = JSON.parse(await zip.file(`json/${version.entityType}.json`).async('text'))
+  const entityUpdates = _.map(({ name, attributes }) => ({
+    entityType: restoredTableName,
+    name,
+    operations: Object.entries(attributes).map(([k, v]) => ({ op: 'AddUpdateAttribute', attributeName: k, addUpdateAttribute: v }))
+  }), entities)
+
+  await Ajax().Workspaces.workspace(namespace, name).upsertEntities(entityUpdates)
+
+  for (const setTableName of _.sortBy(_.identity, version.includedSetEntityTypes)) {
+    const originalSetTableMemberType = setTableName.slice(0, -4)
+
+    const restoredSetTableName = _.replace(version.entityType, restoredTableName, setTableName)
+    const restoredSetTableMemberType = restoredSetTableName.slice(0, -4)
+
+    const setTableEntities = JSON.parse(await zip.file(`json/${setTableName}.json`).async('text'))
+    const setTableEntityUpdates = _.map(({ name, attributes }) => ({
+      entityType: restoredSetTableName,
+      name,
+      operations: Object.entries({
+        ...attributes,
+        [`${originalSetTableMemberType}s`]: _.update(
```

Comment:
I'm confused how this works with the "s" stuck on at the end-- can you briefly explain?

Reply:
This is how Terra names fields in set tables. If there's a "sample" table, the set table is named "sample_set" and contains a column "samples" with a list of references to "sample".